### PR TITLE
perf: PHP to ReactJS controls.

### PR DESCRIPTION
**Pull request by premanshup**
### Customizer - PHP to ReactJS To-Do:

- [x] Check Folder structure and add appropriate Folder Structure.
- [x] Check the Enqueue script for ReactJS - add build to the main customizer control.
- [x] Check the control astra_text_control and remove it at the end.
- [x] Check UI of all the controls and get them fixed.
- [x] Create and combine the Minified existing JS file with new ReactJS build file.
- [x] RTL testing and fix issues ( If any )
- [x] Cross-browser testing and fix issues ( If any )

### Customizer - List of controls:

- [x] require $control_dir . '/class-astra-customizer-control-base.php';
- [x] require $control_dir . '/sortable/class-astra-control-sortable.php';
- [x] require $control_dir . '/radio-image/class-astra-control-radio-image.php';
- [x] require $control_dir . '/slider/class-astra-control-slider.php';
- [x] require $control_dir . '/responsive-slider/class-astra-control-responsive-slider.php';
- [x] require $control_dir . '/responsive/class-astra-control-responsive.php';
- [x] require $control_dir . '/typography/class-astra-control-typography.php';
- [x] require $control_dir . '/responsive-spacing/class-astra-control-responsive-spacing.php';
- [x] require $control_dir . '/divider/class-astra-control-divider.php';
- [x] require $control_dir . '/heading/class-astra-control-heading.php';
- [x] require $control_dir . '/hidden/class-astra-control-hidden.php';
- [x] require $control_dir . '/link/class-astra-control-link.php';
- [x] require $control_dir . '/color/class-astra-control-color.php';
- [x] require $control_dir . '/description/class-astra-control-description.php';
- [x] require $control_dir . '/responsive-color/class-astra-control-responsive-color.php';
- [x] require $control_dir . '/responsive-background/class-astra-control-responsive-background.php';
- [x] require $control_dir . '/border/class-astra-control-border.php';
- [x] require $control_dir . '/customizer-link/class-astra-control-customizer-link.php';
- [x] require $control_dir . '/settings-group/class-astra-control-settings-group.php';
- [x] require $control_dir . '/select/class-astra-control-select.php';
- [x] require $control_dir . '/responsive-select/class-astra-control-responsive-select.php';

### Description

This PR changes the implementation of render function, from PHP to ReactJS for better performance.

### Types of changes

Implementation of the Controls has been changed, all the other logics will remain same.

### How has this been tested?

New React-based controls along with backward compatibility.

### Checklist:
- [x] My code is tested
- [x] My code passes the PHPCS tests
- [x] My code follows accessibility standards 
- [x] My code has proper inline documentation 
- [x] I've included any necessary tests 
- [x] I've included developer documentation 
- [x] I've added proper labels to this pull request

### CODE REVIEW

File: inc/core/class-astra-admin-settings.php [hunk 1]
```diff
@@ -427,8 +427,6 @@ public static function register_scripts( $hook ) {
 				$js_handle[] = 'customize-base';
 			}
```
Comment (imnavanath) [inline inc/core/class-astra-admin-settings.php:428]: We can remove this code from line 416 to 428. No need of this extra variable & LOC.

File: inc/core/common-functions.php [hunk 2]
```diff
@@ -327,16 +327,34 @@ function astra_get_background_obj( $bg_obj ) {
...
-		if ( '' !== $bg_img && '' !== $bg_color ) {
-			$gen_bg_css = array(
-				'background-color' => 'unset',
-				'background-image' => 'linear-gradient(to right, ' . esc_attr( $bg_color ) . ', ' . esc_attr( $bg_color ) . '), url(' . esc_url( $bg_img ) . ')',
-			);
-		} elseif ( '' !== $bg_img ) {
-			$gen_bg_css = array( 'background-image' => 'url(' . esc_url( $bg_img ) . ')' );
-		} elseif ( '' !== $bg_color ) {
-			$gen_bg_css = array( 'background-color' => esc_attr( $bg_color ) );
+			if ( ( 'color' === $bg_type ) ) {
```
Comment (imnavanath) [inline inc/core/common-functions.php:333]: IMO, the switch() case would be better for `$bg_type`.

**Review by vrundakansara — APPROVED** [review]
Looks good to me.

File: inc/customizer/extend-custom-controls/src/border/border-component.js [hunk 3]
```diff
@@ -0,0 +1,150 @@
+import PropTypes from 'prop-types';
+import { Component } from '@wordpress/element';
+import { __ } from '@wordpress/i18n';
+
+class BorderComponent extends Component {
+
+	constructor( props ) {
```
[the remark this answers is not in the document]
Comment (premanshup) [reply inc/customizer/extend-custom-controls/src/border/border-component.js:7]: @shubhw12 Thank you for the suggestion, right now, `@wordpress/scripts` is not allowing us to add Class properties, we will improve this in the next phase when we will convert all the class components to functional components.

File: inc/customizer/extend-custom-controls/src/responsive-spacing/responsive-spacing-component.js [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+
+		if ( undefined === value || '' === value ) {
+			value = this.props.control.params.value;
+		}
+
+		this.state = {
+			value
+		};
+
+		this.renderInputHtml = this.renderInputHtml.bind(this);
```
[the remark this answers is not in the document]
Comment (premanshup) [reply inc/customizer/extend-custom-controls/src/responsive-spacing/responsive-spacing-component.js:21]: @shubhw12 Thank you for the suggestion, right now, `@wordpress/scripts` is not allowing us to add arrow functions, we will improve this in the next phase when we will convert all the class components to functional components.

File: inc/core/common-functions.php (hunk 2 above)
Comment (premanshup) [reply inc/core/common-functions.php:333]: @imnavanath Please verify this, fixed in this PR - #1990

**Review by vrundakansara — APPROVED** [review]

File: inc/core/class-astra-admin-settings.php (hunk 1 above)
Comment (premanshup) [reply inc/core/class-astra-admin-settings.php:429]: @imnavanath I have removed `'wp-color-picker'` dependency as we will still need jQuery. Please verify.

**Review by shubhw12 — APPROVED** [review]

File: inc/core/class-astra-admin-settings.php (hunk 1 above)
Comment (imnavanath) [reply inc/core/class-astra-admin-settings.php:429]: I think some misunderstanding is here, my point was, we don't need this `$js_handle` variable, as it not used anymore in the file so we can remove this code (from line 416 to 428).

Comment (premanshup) [reply inc/core/class-astra-admin-settings.php:429]: @imnavanath Correct!

Comment (premanshup) [reply inc/core/class-astra-admin-settings.php]: @imnavanath Done! 👍🏼

**Review by imnavanath — APPROVED** [review]

**Pull request merged**